src/puppetlabs/puppetdb/http/query.clj

```diff
@@ -396,68 +396,84 @@
+                     ;; entire pg worker.
+                     monitor-deadline (+ deadline 1000000000)
+
+                     ;; May have no response because some tests (e.g. some
+                     ;; with-http-app based tests) don't add one right now.
+                     monitor-id (when-let [chan (and query-monitor
+                                                     (some-> (:response req)
+                                                             response->channel))]
+                                  (qmon/stop-query-at-deadline-or-disconnect query-monitor
+                                                                             query-uuid
+                                                                             chan
+                                                                             monitor-deadline
+                                                                             scf-read-db))]
+                 (try!
+                   [(-> req-with-query-uuid
+                        (assoc :puppetdb-query query-map)
+                        (update :globals merge
+                                {:pretty-print (:pretty query-map pretty-print)
+                                 :query-deadline-ns deadline
+                                 :query-monitor query-monitor}
+                                (when monitor-id
+                                  {:query-monitor-id monitor-id})))
+                    monitor-id]
+                   (catch Exception ex
+                     (when monitor-id (qmon/forget query-monitor monitor-id))
+                     (throw ex))))
+               (catch ExceptionInfo ex
+                 (when-not (= :puppetlabs.puppetdb.query/timeout (:kind (ex-data ex)))
+                   (throw ex))
+                 ;; Note, this will not be reached for the typical
+                 ;; streaming query case because after the query starts, we
+                 ;; return from this function with a 200, and the stream,
+                 ;; and further errors/timeouts have to be handled by the
+                 ;; generated-stream thread feeding jetty.
+                 (let [{:keys [id origin]} (ex-data ex)]
+                   (log/warn (.getMessage ex))
+                   (http/error-response
+                    (if origin
+                      (tru "Query {0} from {1} exceeded timeout" id (pr-str origin))
+                      (tru "Query {0} exceeded timeout" id))
+                    HttpURLConnection/HTTP_INTERNAL_ERROR))))]
+         (if-not mon-id
+           (handler new-req)
+           (try!
+             (let [res (handler new-req)]
+               (when (and mon-id (not (:puppetlabs.puppetdb.query/streaming res)))
+                 (qmon/forget query-monitor mon-id))
+               res)
+             ;; REVIEW: throwable?
+             (catch Exception ex
+               (when mon-id (qmon/forget query-monitor mon-id))
```

Comment:
`mon-id` is always set at this point.

Reply:
Given it should always be forgotten, might be a little easier to follow if the `forget` is moved into a finally block.

Reply:
Is `forget` exception safe?

Reply:
Hmm, I'll have to double-check, but I think this might be the case where we should only `forget` on the error path.  And I believe the `murphy/try!` should handle chaining any pending exception for us.